trieste/models/keras/models.py

```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
+        self.density_estimator.fit(dataset.query_points)
+        self._data_u = self.data_u_appender(x, y)
+
+        xu, yu = self._data_u.astuple()
+
+        # pre-new-oracle, fit u
+        if xu.shape[0] > 0:
+            self.model[1].fit(xu, yu, **self.optimizer.fit_args)
+
+        self._prior_size = dataset.query_points.shape[0]
+        self.optimizer.optimizer.learning_rate.assign(self._learning_rate)
+
+    def predict(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        r"""
+        Returns mean and variance at ``query_points`` for the Direct Epistemic
+        Uncertainty Prediction model.
+
+        Following <cite data-cite="jain2022"/>, we consider the case for a model with
+        a Gaussian ground truth, i.e.
+                .. math:: p(y \mid x)=\mathcal{N}\left(y ; f^{*}(x), \sigma^{2}(x)\right).
+
+        In this scenario, it can be shown that the epistemic uncertainty can be estimated
+        as the squared difference between main-model predictions and the Bayes optimal
+        prediction at a given point, that is:
+                .. math:: \mathcal{E}(\hat{f}, x)=\left(\hat{f}(x)-f^{*}(x)\right)^{2}
+
+        As the Bayes optimal is unattainable, we instead approximate the epistemic uncertainty
+        by computing the total uncertainty of the model, which is captured by the expected
+        squared loss of the main-model predictions. This can be deconstructed as follows
+                .. math:: \mathcal{U}(\hat{f}, x)=E_{P(. \mid x)}\left[(\hat{f}(x)-y)^{2}\right]=\left(\hat{f}(x)-f^{*}(x)\right)^{2}+\sigma^{2}(x)
+        
+        The predict method returns the point estimates of the main predictor, and uses point estimates
+        of the squared loss to return predictions for model uncertainty. Note that the dataset used to
+        train the models in a BO setting grows over time, which renders estimates of uncertainty
+        as iteration dependent. Following the original paper, we account for this evolving uncertainty 
+        space by considering stationarizing variables, namely the predicted variance from the ensemble 
+        model and the Gaussian density of our queried points. These are used to enhance the dataset of
+        features the auxiliary model predicts on, so that a D-dimensional queried point will be passed
+        as a [D+2,1] observation to the error predictor.
+
+        :param query_points: The points at which to make predictions.
+        :return: The predicted mean and variance of the observations at the specified
+            ``query_points``.
+        """
+        if not tf.is_tensor(query_points):
+            query_points = tf.convert_to_tensor(query_points)
+        if query_points.shape.rank == 1:
+            query_points = tf.expand_dims(query_points, axis=-1)
+
+        f_pred, f_var = self.model[0].predict(query_points)
+        f_pred, f_var = tf.cast(f_pred, dtype=tf.float64), tf.cast(f_var, dtype=tf.float64)
```

Comment:
Surprisingly, no. The data in _get_example_data is float64, and the output predictions (as standalone or as main predictor here) of the deep ensembles are float32. Since they are not concatenated this is not flagged as an issue, but with the tf.concats here I need to homogenize dtypes.

Reply:
Ok, so I did a general fix. Now I infer the data dtype, and ensure variance and density use the same (this required quite a few changes in the density class, but it works). I included a test that validates the changes. Thanks for bringing it up :)